application/src/main/java/org/thingsboard/server/service/edge/rpc/processor/user/BaseUserProcessor.java

```diff
@@ -0,0 +1,128 @@
+
+    protected boolean saveOrUpdateUser(TenantId tenantId, UserId userId, UserUpdateMsg userUpdateMsg) {
+        boolean isCreated = false;
+
+        try {
+            User user = JacksonUtil.fromString(userUpdateMsg.getEntity(), User.class, true);
+            if (user == null) {
+                throw new RuntimeException("[{" + tenantId + "}] userUpdateMsg {" + userUpdateMsg + "} cannot be converted to user");
+            }
+
+            User userById = edgeCtx.getUserService().findUserById(tenantId, userId);
+            if (userById == null) {
+                isCreated = true;
+                user.setId(null);
+            } else {
+                user.setId(userId);
+            }
+            setCustomerId(isCreated ? null : userById.getCustomerId(), user);
+
+            userValidator.validate(user, User::getTenantId);
+
+            if (isCreated) {
+                user.setId(userId);
+                String email = user.getEmail();
+                User userByEmail = edgeCtx.getUserService().findUserByEmail(tenantId, email);
+                if (userByEmail != null && !userByEmail.getId().equals(userId)) {
+                    throw new DataValidationException(String.format(
+                            "[%s] User with email %s already exists!", tenantId, email
+                    ));
+                }
+            }
+
+            edgeCtx.getUserService().saveUser(tenantId, user, false);
+        } catch (Exception e) {
+            log.error("[{}] Failed to process user update msg [{}]", tenantId, userUpdateMsg, e);
+            throw e;
+        }
+
+        return isCreated;
+    }
+
+    protected void updateUserCredentials(TenantId tenantId, UserCredentialsUpdateMsg userCredentialsUpdateMsg) {
+        UserCredentials userCredentials = JacksonUtil.fromString(userCredentialsUpdateMsg.getEntity(), UserCredentials.class, true);
+        if (userCredentials == null) {
+            throw new RuntimeException("[{" + tenantId + "}] userCredentialsUpdateMsg {" + userCredentialsUpdateMsg + "} cannot be converted to user credentials");
+        }
+        User user = edgeCtx.getUserService().findUserById(tenantId, userCredentials.getUserId());
+        if (user != null) {
+            log.debug("[{}] Updating user credentials for user [{}]. New credentials Id [{}], enabled [{}]",
+                    tenantId, user.getName(), userCredentials.getId(), userCredentials.isEnabled());
+            try {
+                UserCredentials userCredentialsByUserId = edgeCtx.getUserService().findUserCredentialsByUserId(tenantId, user.getId());
+
+                if (userCredentialsByUserId == null) {
+                    userCredentialsByUserId = new UserCredentials();
+                }
+
+                userCredentialsByUserId.setId(userCredentials.getId());
+                userCredentialsByUserId.setEnabled(userCredentials.isEnabled());
+                userCredentialsByUserId.setActivateToken(userCredentials.getActivateToken());
```

Comment:
Maybe this can be moved as well? 
I suppose it can be moved into 'if (userCredentialsByUserId == null)' block